File: app/channels/research_solution_channel.rb
class ResearchSolutionChannel < ApplicationCable::Channel
  def subscribed
    stream_for solution
  end

  def unsubscribed
  end

  def create_submission(data)
    uuid = SecureRandom.uuid

    SubmissionServices::Create.(uuid, solution, data["submission"])
  end

  private

  def solution
    @solution ||= Research::ExperimentSolution.find(params[:id])
  end
end

File: app/javascript/channels/research_solution_channel.js
import consumer from "./consumer"

class ResearchSolutionChannel {
  constructor(id) {
    this.id = id;
    this.subscription = consumer.subscriptions.create(
      { channel: "ResearchSolutionChannel", id: this.id }
    );
  }

  received(handler) {
    this.subscription.received = handler;
  }

  createSubmission(submission) {
    this.subscription.perform("create_submission", { submission: submission });
  }

  cancelSubmission() {
  }
}

export default ResearchSolutionChannel;
